src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ApiCompat.Common.targets

```diff
@@ -10,18 +10,24 @@ Copyright (c) .NET Foundation. All rights reserved.
 ***********************************************************************************************
 -->
 <Project>
+
   <!-- Expose the tasks for SDK consumption and for external use cases. -->
   <UsingTask TaskName="Microsoft.DotNet.ApiCompat.Task.ValidateAssembliesTask" AssemblyFile="$(DotNetApiCompatTaskAssembly)" />
   <UsingTask TaskName="Microsoft.DotNet.ApiCompat.Task.ValidatePackageTask" AssemblyFile="$(DotNetApiCompatTaskAssembly)" />
 
   <Target Name="CollectApiCompatInputs">
     <PropertyGroup Condition="'$(RoslynAssembliesPath)' == ''">
       <RoslynAssembliesPath>$(RoslynTargetsPath)</RoslynAssembliesPath>
-      <_packageReferenceList>@(PackageReference)</_packageReferenceList>
+
+      <_UsesRoslynToolsetPackage Condition="'@(PackageReference->AnyHaveMetadataValue('Identity', 'Microsoft.Net.Compilers.Toolset'))' == 'true'">true</_UsesRoslynToolsetPackage>
       <!-- CSharpCoreTargetsPath and VisualBasicCoreTargetsPath point to the same location, Microsoft.CodeAnalysis.CSharp and Microsoft.CodeAnalysis.VisualBasic
            are on the same directory as Microsoft.CodeAnalysis. So there is no need to distinguish between csproj or vbproj. -->
-      <RoslynAssembliesPath Condition="$(_packageReferenceList.Contains('Microsoft.Net.Compilers.Toolset'))">$([System.IO.Path]::GetDirectoryName($(CSharpCoreTargetsPath)))</RoslynAssembliesPath>
+      <RoslynAssembliesPath Condition="'$(_UsesRoslynToolsetPackage)' == 'true'">$([System.IO.Path]::GetDirectoryName($(CSharpCoreTargetsPath)))</RoslynAssembliesPath>
       <RoslynAssembliesPath Condition="'$(MSBuildRuntimeType)' == 'Core'">$([System.IO.Path]::Combine('$(RoslynAssembliesPath)', bincore))</RoslynAssembliesPath>
+
+      <!-- Temporary workaround to mitigate https://github.com/dotnet/sdk/issues/32691 with roslyn toolset package and desktop msbuild < 17.9.0.
+           Reset the roslyn assemblies path back to Roslyn inside Visual Studio. -->
+      <RoslynAssembliesPath Condition="'$(MSBuildRuntimeType)' != 'Core' and '$(_UsesRoslynToolsetPackage)' == 'true' and '$(MSBuildVersion)' &lt; '17.9.0'">$(RoslynTargetsPath)</RoslynAssembliesPath>
```

Comment:
For some reason I thought we were also hitting this in 8.0.2xx when not using the compilers package, is that true?

Reply:
If the app isn't using the compiler toolset package then that means that msbuild and roslyn versions are in sync. At least that's my understanding.

Reply:
> Just a thought on how automatic the consumption of Microsoft.Net.Compilers.Toolset should be -- what if we just flipped the default?

I'm not sure I would want to make such a change as part of the 8.0.1xx servicing release (i.e. for source-build SDKs) and for the 8.0.2xx feature branch. But I'm inclined to consider that for main. Would we then enable that setting in runtime? We would still need the temporary workaround.

Reply:
We could enable it in runtime - though we might not *need* to and would be OK to disable it if it caused a problem.  It's rare that we *need* a newer compiler in APICompat - we usually just *want* one to pick up a bugfix.

So the suggestion would be to just change this to do:
```xml
<RoslynAssembliesPath Condition="'$(_UsesRoslynToolsetPackage)' == 'true' AND '$(ApiCompatUseRoslynToolsetPackagePath)' == 'true'">$([System.IO.Path]::GetDirectoryName($(CSharpCoreTargetsPath)))</RoslynAssembliesPath>
```

I'm OK if you do that in `main` or both `main` and `8.0`.

Reply:
I'm OK with applying this in both main and .NET 8.0.1xx. Updated the PR. Please take a look.

> I know the compiler team doesn't think of this toolset package as supported, so it's a bit unusual that we automatically pick it up.

I was surprised by that as well when I started working in this codebase.

Reply:
I also submitted https://github.com/dotnet/arcade/pull/14459 to update GenFacades which uses the same pattern.

Reply:
> I was surprised by that as well when I started working in this codebase.

I think when we initially started working on this we talked about not wanting to ship the compilers and instead locate the ones used by the project.  Naturally folks discovered that "used by the project" sometimes meant from a nuget package - and I don't think we had a gauge for what's normal vs exceptional.